sandy/core/cov.py

```diff
@@ -2480,6 +2301,92 @@ def sample_distribution(dim, nsmp, seed=None, pdf='normal'):
     return y
 
 
+def reduce_size(data):
+    """
+    Reduces the size of the matrix, erasing the zero values.
+
+    Parameters
+    ----------
+    data : 'pd.DataFrame'
+        Matrix to be reduced.
+
+    Returns
+    -------
+    nonzero_idxs : `numpy.ndarray`
+        The indices of the diagonal that are not null.
+    cov_reduced : `pandas.DataFrame`
+        The reduced matrix.
+
+    Examples
+    --------
+    >>> S = pd.DataFrame(np.diag(np.array([1, 2, 3])))
+    >>> non_zero_index, reduce_matrix = reduce_size(S)
+    >>> assert reduce_matrix.equals(S)
+    >>> assert (non_zero_index == range(3)).all()
+
+    >>> S = pd.DataFrame(np.diag(np.array([0, 2, 3])))
+    >>> non_zero_index, reduce_matrix = reduce_size(S)
+    >>> assert (non_zero_index == np.array([1, 2])).all()
+    >>> reduce_matrix
+      1 2
+    1 2 0
+    2 0 3
+
+    >>> S.index = S.columns = ["a", "b", "c"]
+    >>> non_zero_index, reduce_matrix = reduce_size(S)
+    >>> reduce_matrix
+      b c
+    b 2 0
+    c 0 3
+    """
+    data_ = pd.DataFrame(data)
+    nonzero_idxs = np.flatnonzero(np.diag(data_))
+    cov_reduced = data_.iloc[nonzero_idxs, nonzero_idxs]
+    return nonzero_idxs, cov_reduced
+
+
+def restore_size(nonzero_idxs, mat_reduced, dim):
+    """
+    Restore the size of a matrix.
+
+    Parameters
+    ----------
+    nonzero_idxs : `numpy.ndarray`
+        The indices of the diagonal that are not null.
+    mat_reduced : `numpy.ndarray`
+        The reduced matrix.
+    dim : `int`
+        Dimension of the original matrix.
+
+    Returns
+    -------
+    mat : `pd.DataFrame`
+        Matrix of specified dimensions.
+
+    Notes
+    -----
+    ..notes:: This funtion was developed to be used after using
+              `reduce_size`.
+
+    Examples
+    --------
+    >>> S = pd.DataFrame(np.diag(np.array([0, 2, 3, 0])))
+    >>> M_nonzero_idxs, M_reduce = reduce_size(S)
+    >>> M_reduce[::] = 1
+    >>> restore_size(M_nonzero_idxs, M_reduce.values, len(S))
+                0           1           2           3
+    0 0.00000e+00 0.00000e+00 0.00000e+00 0.00000e+00
+    1 0.00000e+00 1.00000e+00 1.00000e+00 0.00000e+00
+    2 0.00000e+00 1.00000e+00 1.00000e+00 0.00000e+00
+    3 0.00000e+00 0.00000e+00 0.00000e+00 0.00000e+00

```

Comment:
Can we check that the combination `reduce_size` and then `restore_size ` preserves indices and columns?

Reply:
I put a test demostrating that the original index are not kept in the restore part. For cov reduced is kept.